### ISSUE TITLE
[Pocket] Add telemetry to Pocket video feed

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#777]
Goal + KPI
Confirm hypotheses that by adding Pocket recommended video feed, people will interact with the feature (click on the mega tile), and find it interesting enough to click, watch, and stay longer in app.

Be able to identify users who use the feature, and bucket them to see if their session time is higher, and retention is better (come back more often)

**Add additional probes**

Megatile:
- [x] Click on mega tile

Feed list:
- [ ] Probe to see click-through-rate of each video (Pocket ask) per client
- [x] Probe to get number of videos played per session/per user

**Main KPI**
- DAU
- Retention (2 points, 7 day, in comparison to April 30 data point) (We believe retention of Pocket feed viewers is higher than non-Pocket)
- Positive feedback on Amazon store
- Increase of app rating from 4.2. to 4.3

**Who is a Pocket user?**
- A user who is active for x days and has watched x videos (@mmccorks to confirm once data is coming in)

**Future**
- Funnel, e.g. what does a user do before/after using feature
- click-through-rate for Pocket videos

**Other questions**
- Did they just click on it or actually stayed and finished watching it
- Can people remove the tile?  (I don't think so)

**Comment by kennylong** [comment]
To be consistent with how we are measuring recommendations in the Firefox new tab, Pocket would need impressions and clicks per video recommendation (unique `id` provided in feed response).

**Comment by mcomella** [comment]
> Pocket would need impressions and clicks per video recommendation

afaik, collecting this would go against our data policies (because we can correlate visited URLs with specific users). Can we click use the pocket redirect links in the video feed and pass you an identifier, e.g. `?src=firefox-tv` so we can measure this data server-side where we can't easily correlate specific users to URLs?

@liuche Is this correct? If so, please at-mention Kenny to see if this is a valid solution.

**Comment by bbinto** [comment]
@mcomella besides clarifying Kenny's last comment, are you ok with the goals, probes I've layed out here?

**Comment by kennylong** [comment]
@mcomella For the new tab recommendations impression/click analytics we encountered this same data collection issue, and the agreed-upon implementation was:

(1) In Fx 57 (initial launch), we excluded the client_id for clicks, so it could not be tied to a user.

This wasn't ideal for a few reasons (e.g. couldn't count uniques), so we changed to this:

(2) In Fx 58+, we used a "pocket-id" in place of the client_id for all data. The pocket-id is basically just a unique identifier that isn't used anywhere else and is in no way tied to a client_id.

Could we do something like #2? We could get more information from the activity stream team about how they implemented the "pocket-id".

**Comment by liuche** [comment]
Emailed Legal to ask about this, we may need to change our Privacy Policy to include this (especially since we're still using Focus's, which we won't want to include this under), which is #692 .

**Comment by mcomella** [comment]
> Probe to see click-through-rate of each video (Pocket ask)

@bbinto Two questions:
- Should we have a probe for each time a video is clicked or just one at the end with the count?
- Should this be the number of *unique* videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.

**Comment by mcomella** [comment]
> Confirming with team about this,

afaik, we're still waiting. @liuche Any updates?

**Comment by liuche** [comment]
Have a meeting scheduled for Wednesday, I'll update after that.

**Comment by bbinto** [comment]
@kennylong to confirm please, see @mcomella 

- Should we have a probe for each time a video is clicked or just one at the end with the count?
- Should this be the number of unique videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.

When you say "at the end", is that when the sessions has ended / app is in background?

**Comment by liuche** [comment]
Okay, it looks like the proposed data collection here is:

<pocket video url>: number of clicks

so just to be explicit @kennylong, the data collected by our local telemetry would be something like:

```
{ client-id: a874ebfeef-...:
    { https://pocket.co/xQDbu8: 5,
      https://pocket.co/bXTno3: 1
    },
   ...
}
```

where client-id is a per-client randomly generated id, not tied to anything else.

Is this what is being proposed as collection? If so, give me a few days to make sure that's okay.
If not, please update with a similar code snippet.

**Comment by bbinto** [comment]
@kennylong to confirm please, see @mcomella 

- Should we have a probe for each time a video is clicked or just one at the end with the count?
- Should this be the number of unique videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.

When you say "at the end", is that when the sessions has ended / app is in background?

**Comment by kennylong** [comment]
@liuche @bbinto Sorry for the delay on this. The notification emails aren't going to my work email for some reason. Will get that fixed.

I'm not actually sure about the format of the raw data since my only experience so far is just consuming data that's in the `tiles` database. Do we need to loop in someone else for this? Or we can go the route of having us access the raw data (in which case we'll need to figure out how we're going to do that and also we might have some suggestions on format).

Some things I can weigh in on:
- I think we'd prefer things to be logged with the `id` rather than the `url`
- Randomly generated client-id is fine, but will persist?
- Total clicks per client per pocket-video-url is great

If you want to discuss this in more detail than this thread feel free to Slack me or set up a call.

**Comment by liuche** [comment]
@bbinto who should we loop in for passing the data over? or can they use the Telemetry dashboard?

That aside, I think this data structure is clear enough, and we can move to data review whenever it's ready.

**Comment by mcomella** [comment]
@liuche If we use our standard telemetry system, won't it include telemetry user UUIDs which would let us correlate users to videos clicked and make this a no-go? I'd think we have to create a new ping (will that work?), or use a separate system, that would only include a Pocket ID.

**Comment by mcomella** [comment]
Sorry, I quoted the wrong text with my question in https://github.com/mozilla-mobile/firefox-tv/issues/777#issuecomment-385829411.

> Probe to get number of videos played per session/per user

@bbinto Two questions:
- Should we have a probe for each time a video is clicked or just one at the end with the count?
- Should this be the number of *unique* videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.

**Comment by bbinto** [comment]
@mcomella, I favour not including a new ping to our telemetry system, and rather use what we have and will pass the data review, as well as giving Pocket folks access to our redash queries/dashboard.

> Probe to get number of videos played per session/per user

- Should we have a probe for each time a video is clicked or just one at the end with the count?
At the end is fine

- Should this be the number of unique videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.
How much more difficult is the former?

**Comment by kennylong** [comment]
@bbinto A question regarding this comment:

> I favour not including a new ping to our telemetry system, and rather use what we have and will pass the data review, as well as giving Pocket folks access to our redash queries/dashboard.

Being able to query in Redash to get clicks per video id per day would be work for us (Pocket) from a data flow/access perspective. Is that what this would enable?

Separately, I'm realizing that impressions of video thumbnails (by video id) is not included in this. Not sure if it was missed [in my original comment](https://github.com/mozilla-mobile/firefox-tv/issues/777#issuecomment-381207297) or if it was decided against, but given that impressions will not be even throughout the day, this is an important part of allowing us to measure the overall user-interest (by CTR) of each video, which is how we plan to improve what we are adding to the video feed. If we only have click data to rely on, it will be tough to tell if a video got clicks because it was interesting or if it was because it had prominent placement during a high-visibility time of the day.

**Comment by mcomella** [comment]
> Should this be the number of unique videos clicked or would it be fine to count the total number of clicks? The latter is simpler for us.
How much more difficult is the former?

Now that I think about it, since it's per session and we don't need to persist beyond app runs, not very. I'll do the former.

**Comment by mcomella** [comment]
FYI: opened a PR for the probes that are not click-through rate (which is blocked on discussions).

**Comment by liuche** [comment]
Thanks for the initial telemetry PR, I'm still following up on the pocket video id impressions telemetry.

**Comment by liuche** [comment]
@mcomella here's the [telemetry](https://github.com/mozilla/activity-stream/blob/master/docs/v2-system-addon/data_dictionary.md#example-activity-stream-impression_stats-logs) that Activity Stream sends. They use pingcenter for AS data, and can send pings whenever events occur (without aggregating or waiting) that is not connected to UI Telemetry or coreping.

The key takeaways are:
* This is a separate ping from UI Telemetry or coreping, and *must not* include clientid is used by Firefox Telemetry
* Uses a impression_id which is unique to each client, but not connected to client id and used for nothing else

So it looks like for this bug, we'll have to use a separate ping, to be consistent with what AS does.

**Comment by mcomella** [comment]
> I favour not including a new ping to our telemetry system,

@bbinto Given Chenxia's most recent comment, how do we want to handle this? Should we add a new ping type? Should we talk to Pocket to see if we can simplify by tracking this on the back-end? Is it necessary to collect this data for v3.0 (I assume yes)?

**Comment by bbinto** [comment]
Sorry for the delay, @mcomella.

So the only outstanding probe that needs to be decided on is the click-through probe which would need a new ping type, right?

Do we have an idea how much effort this is for our team, and our data team? Not sure if they have time to help us with that.

**Comment by mcomella** [comment]
@bbinto Right. If the server is written to generically accept new ping types (which i think it might be nowadays), I'd say M. If not, it'd take some back-and-forth with the data team so L or larger.

**Comment by liuche** [comment]
@ekager recently did this for Focus, so we might be able to share this code.

**Comment by mcomella** [comment]
@ekager How long did it take you to add a new ping type to Focus?

Size L to be safe because we're not sure how much work this would be.

**Comment by ekager** [comment]
Definitely at least a M but probably a L because you need to update the Telem library with the new ping type, wait for the next release of AC, implement the new ping type in the app, and then validate (and maybe debug) with the data team. Overall took me ~5-6 days but over 2-3 weeks, and I could probably do it faster now that I know what’s going on.

**Comment by bbinto** [comment]
**Main KPI**
- DAU
- Retention (2 points, 7 day, in comparison to April 30 data point)
- Positive feedback on Amazon store
- Increase of app rating from 4.2. to 4.3

**Comment by bbinto** [comment]
> @ekager recently did this for Focus, so we might be able to share this code.

@ekager what was the reasoning for that, and what ping is it?

**Comment by ekager** [comment]
@bbinto mobile-metrics ping for GV histogram data

**Comment by mcomella** [comment]
Not actively working on this: looks like I landed my WIP on master in #881.

**Comment by Sdaswani** [comment]
I will talk to Matt Koidin today about this @liuche but can you keep it a ‘nice to have’ for 3.0 to see if we can get to it otherwise we can punt to 3.1.

**Comment by Sdaswani** [comment]
@kennylong can you advise us on this - is it ok if we get the new telemetry in 3.1 instead of 3.0? We would lose a few weeks of data but it is the first few weeks of that data so may be a bigger miss than it seems.

**Comment by kennylong** [comment]
@Sdaswani It won't cause a problem on our end to wait for 3.1 (assuming that's not too far off in the future).

**Comment by Sdaswani** [comment]
@liuche please put this in the 3.1 sprint, and thanks @kennylong !

**Comment by athomasmoz** [comment]
@psymoon What's the status of this one? Is it blocked on a-c?

**Comment by psymoon** [comment]
@athomasmoz I'm working on this on the a-c side

**Comment by athomasmoz** [comment]
@psymoon Is there a list of the new telemetry probes being added somewhere?

**Comment by psymoon** [comment]
@athomasmoz No, where do they generally added?

**Comment by athomasmoz** [comment]
Hmm I think they're typically added here: https://github.com/mozilla-mobile/firefox-tv/blob/master/docs/telemetry.md

**Comment by mcomella** [comment]
Ashley is correct: please document all new probes in the telemetry documentation in-tree.

**Comment by pocmo** [comment]
Removing a-c label: @psymoon has landed the needed code in AC. To my knowledge no additional changes in AC are required.

**Comment by psymoon** [comment]
Needs https://github.com/mozilla-mobile/android-components/pull/2000 merge and a-c upgrade thereafter
